# Data cleaning before model training

In [ ]:
# Mount Google Drive
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

In [ ]:
# Load dataset
csv_path = "/content/drive/MyDrive/Data/alldata_1_for_kaggle.csv"
columns = ["Serial Number","Class Labels","Research Paper Text"]
df = pd.read_csv(csv_path, header=None, encoding = "latin-1", names=columns, skiprows=[0])

In [ ]:
df

Serial Number    Class Labels  \
0                 0  Thyroid_Cancer   
1                 1  Thyroid_Cancer   
2                 2  Thyroid_Cancer   
3                 3  Thyroid_Cancer   
4                 4  Thyroid_Cancer   
...             ...             ...   
7565           7565    Colon_Cancer   
7566           7566    Colon_Cancer   
7567           7567    Colon_Cancer   
7568           7568    Colon_Cancer   
7569           7569    Colon_Cancer   

                                    Research Paper Text  
0     Thyroid surgery in  children in a single insti...  
1     " The adopted strategy was the same as that us...  
2     coronary arterybypass grafting thrombosis fib...  
3      Solitary plasmacytoma SP of the skull is an u...  
4      This study aimed to investigate serum matrix ...  
...                                                 ...  
7565  we report the case of a 24yearold man who pres...  
7566  among synchronous colorectal cancers scrcs rep...  
7567  the heterogeneity of cancer cells is generally...  
7568  "adipogenesis is the process through which mes...  
7569  the periparturient period is one of the most c...  

[7570 rows x 3 columns]

In [ ]:
len(df)

7570

In [ ]:
# Show unique values
df["Class Labels"].drop_duplicates()

0      Thyroid_Cancer
281      Colon_Cancer
539       Lung_Cancer
Name: Class Labels, dtype: object

In [ ]:
df["Research Paper Text"].drop_duplicates()

0       Thyroid surgery in  children in a single insti...
1       " The adopted strategy was the same as that us...
2       coronary arterybypass grafting thrombosis fib...
3        Solitary plasmacytoma SP of the skull is an u...
4        This study aimed to investigate serum matrix ...
                              ...                        
6863    "Missense mutation distribution in the exons a...
6929    "versus gemcitabine/carboplatin in advanced no...
7040     Keloids are pathological scars that grow over...
7485    the anization of cells into multiple membranou...
7497    several immunotherapeutic strategies that harn...
Name: Research Paper Text, Length: 996, dtype: object

In [ ]:
# Drop duplicates
df = df.drop_duplicates(subset=["Research Paper Text"])

In [ ]:
len(df)

996

In [ ]:
df

Serial Number    Class Labels  \
0                 0  Thyroid_Cancer   
1                 1  Thyroid_Cancer   
2                 2  Thyroid_Cancer   
3                 3  Thyroid_Cancer   
4                 4  Thyroid_Cancer   
...             ...             ...   
6863           6863     Lung_Cancer   
6929           6929     Lung_Cancer   
7040           7040  Thyroid_Cancer   
7485           7485    Colon_Cancer   
7497           7497    Colon_Cancer   

                                    Research Paper Text  
0     Thyroid surgery in  children in a single insti...  
1     " The adopted strategy was the same as that us...  
2     coronary arterybypass grafting thrombosis fib...  
3      Solitary plasmacytoma SP of the skull is an u...  
4      This study aimed to investigate serum matrix ...  
...                                                 ...  
6863  "Missense mutation distribution in the exons a...  
6929  "versus gemcitabine/carboplatin in advanced no...  
7040   Keloids are pathological scars that grow over...  
7485  the anization of cells into multiple membranou...  
7497  several immunotherapeutic strategies that harn...  

[996 rows x 3 columns]

In [ ]:
df["Research Paper Text"][0]

'thyroid surgery child single institution osama ibrahim almosallama ali aseerib ahmed alhumaida ali alzahranic saif alsobhib saud alshanafeybfrom adepartment surgery college medicine qassim university buraidah al qassim saudi arabia bdepartment surgery king faisal specialist hospital research center riyadh saudi arabia cdepartment medicine king faisal specialist hospital research center riyadh saudi arabia correspondence dr osama ibrahim almosallam department surgery college medicine qassim university po box buraidah al qassim saudi arabia osama_iaahotmailcom orcid orcid0000000290367564 citation almosallam oi aseeri alhumaid alzahrani alsobhi alshanafey thyroid surgery child single institution ann saudi med received january accepted may published august copyright copyright annals saudi medicine saudi arabia access creative common attributionnoncommercialnoderivatives international license cc byncnd detail accessed httpcreativecommons licensesbyncnd40funding nonebackground data thyroid 

By dropping duplicates we can go from 7570 values to 996 unique values.

## Data preprocessing functions

In [ ]:
!pip install contractions unidecode nltk

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.5/235.5 kB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 289.9/289.9 kB 10.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.8/110.8 kB 13.4 MB/s eta 0:00:00


In [ ]:
import re
from contractions import fix
import nltk
from unidecode import unidecode
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
def toLowerCase(data):
  return data.lower()

In [ ]:
def removeHyperlinksAndHtmlTags(data):
    data = re.sub('<.*?>+', '', data)
    data = re.sub('(https|http)?://\S+|www\.\S+', '', data)
    return data

In [ ]:
def removePunctuation(data):
  return re.sub(r'[^\w\s]+', '', data)

In [ ]:
def removeNewlinesAndTabs(data):
  return data.replace("\\n"," ").replace("\t"," ")

In [ ]:
# Fixes contractions such as `you're` to you `are`
def fixesContractions(data):
  return fix(data)

In [ ]:
# handling accented character
def handleAccentedCharacters(data):
  return unidecode(data)

In [ ]:
# remove stopwords such as i, me, my, myself, ...
def removeStopwords(data):
    stopwords_list = set(stopwords.words('english'))

    words = word_tokenize(data)

    filtered_words = [word for word in words if word.lower() not in stopwords_list]

    filtered_text = ' '.join(filtered_words)

    return filtered_text

In [ ]:
# defining the function for getting root
def lemmatization(data):
    lem = WordNetLemmatizer()
    lem_words = [lem.lemmatize(word) for word in data.split()]
    return " ".join(lem_words)

##  Clean dataset

In [ ]:
df['Research Paper Text'] = df['Research Paper Text'].apply(toLowerCase)
df['Research Paper Text'] = df['Research Paper Text'].apply(removeHyperlinksAndHtmlTags)
df['Research Paper Text'] = df['Research Paper Text'].apply(removePunctuation)
df['Research Paper Text'] = df['Research Paper Text'].apply(removeNewlinesAndTabs)
df['Research Paper Text'] = df['Research Paper Text'].apply(fixesContractions)
df['Research Paper Text'] = df['Research Paper Text'].apply(handleAccentedCharacters)
df['Research Paper Text'] = df['Research Paper Text'].apply(removeStopwords)
df['Research Paper Text'] = df['Research Paper Text'].apply(lemmatization)

<ipython-input-21-68dd65a0d1ea>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Research Paper Text'] = df['Research Paper Text'].apply(toLowerCase)
<ipython-input-21-68dd65a0d1ea>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Research Paper Text'] = df['Research Paper Text'].apply(removeHyperlinksAndHtmlTags)
<ipython-input-21-68dd65a0d1ea>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See 

In [ ]:
df

Serial Number    Class Labels  \
0                 0  Thyroid_Cancer   
1                 1  Thyroid_Cancer   
2                 2  Thyroid_Cancer   
3                 3  Thyroid_Cancer   
4                 4  Thyroid_Cancer   
...             ...             ...   
6863           6863     Lung_Cancer   
6929           6929     Lung_Cancer   
7040           7040  Thyroid_Cancer   
7485           7485    Colon_Cancer   
7497           7497    Colon_Cancer   

                                    Research Paper Text  
0     thyroid surgery child single institution osama...  
1     adopted strategy used prior year based four ex...  
2     coronary arterybypass grafting thrombosis ibri...  
3     solitary plasmacytoma sp skull uncommon clinic...  
4     study aimed investigate serum matrix metallopr...  
...                                                 ...  
6863  missense mutation distribution exon functional...  
6929  versus gemcitabinecarboplatin advanced nonsmal...  
7040  keloid pathological scar grow time extend beyo...  
7485  anization cell multiple membranous compartment...  
7497  several immunotherapeutic strategy harness exq...  

[996 rows x 3 columns]

## Create a new CSV for the cleaned dataset

In [ ]:
cleaned_dataset= df.to_csv('/content/drive/MyDrive/Data/CleanedMedicalDataset.csv')